File: lib/jump.js
'use strict';

// Jump to the window a session lives in (NOTES.md, Phase 5: jump to session).
// src-tauri/src/jump.rs is a port of this file; the two must agree.
//
// Desktop-app sessions: the Claude app handles claude://code/continue?session=<id>,
// which opens an existing session without creating one. Undocumented (it's what the
// app builds for its own Spotlight results), so it can break with an app update, and
// it needs the app's own session id, found in its session store by our CLI id.
// Never claude-cli://open?cwd=...: that starts a new session.
//
// Terminal sessions: the session's process has a tty. In Terminal.app the tab with
// that tty is selected and brought forward; in any other app, the app is activated.

const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');

const LOCAL_ID_RE = /^local_[A-Za-z0-9-]{1,64}$/; // the Claude app's own validation
const TTY_RE = /^ttys\d{1,4}$/;
const DESKTOP_BUNDLE_ID = 'com.anthropic.claudefordesktop';
const MAX_ANCESTORS = 15;

function desktopStore(home) {
  return path.join(home, 'Library', 'Application Support', 'Claude', 'claude-code-sessions');
}

function coworkStore(home) {
  return path.join(home, 'Library', 'Application Support', 'Claude', 'local-agent-mode-sessions');
}

// The Claude app's record of a session, found by our (CLI) session id. The store is
// <account>/<org>/local_<id>.json; files are skipped unless they mention the id.
function desktopSession(home, cliSessionId, root = desktopStore(home)) {
  const dirs = (dir) => {
    try {
      return fs.readdirSync(dir, { withFileTypes: true }).filter((d) => d.isDirectory()).map((d) => path.join(dir, d.name));
    } catch {
      return [];
    }
  };
  for (const org of dirs(root).flatMap(dirs)) {
    let names;
    try {
      names = fs.readdirSync(org);
    } catch {
      continue;
    }
    for (const name of names) {
      if (!name.startsWith('local_') || !name.endsWith('.json')) continue;
      let raw;
      try {
        raw = fs.readFileSync(path.join(org, name), 'utf8');
      } catch {
        continue;
      }
      if (!raw.includes(cliSessionId)) continue;
      let record;
      try {
        record = JSON.parse(raw);
      } catch {
        continue;
      }
      if (record && record.cliSessionId === cliSessionId && LOCAL_ID_RE.test(record.sessionId)) {
        return { localId: record.sessionId, archived: Boolean(record.isArchived) };
      }
    }
  }
  return null;
}

// One `ps -o ppid=,tty=,comm=` line.
function parsePs(line) {
  const m = /^\s*(\d+)\s+(\S+)\s+(.+?)\s*$/.exec(String(line || ''));
  return m ? { ppid: Number(m[1]), tty: m[2], command: m[3] } : null;
}

function psInfo(pid) {
  try {
    return parsePs(execFileSync('/bin/ps', ['-o', 'ppid=,tty=,comm=', '-p', String(pid)], { encoding: 'utf8', timeout: 3000 }));
  } catch {
    return null;
  }
}

// "/Applications/X.app/Contents/MacOS/X" -> "/Applications/X.app". Claude Code's own
// binary can live inside a claude.app bundle, which isn't the window we want.
function appBundle(command) {
  const m = /^(.*?\.app)\/Contents\/MacOS\//.exec(String(command || ''));
  if (!m || /\/claude\.app$/.test(m[1])) return null;
  return m[1];
}

// What a jump to this session would do, or why it can't. ps is injectable for tests.
function planJump(session, { home, ps = psInfo }) {
  if (!session || typeof session.session_id !== 'string') return { status: 404, error: 'no such session' };

  // Cowork has no link to open an existing task, but the app's "waiting for you" link
  // opens a Cowork session while it waits on a permission, the case that matters.
  // Any other Cowork state just brings the Claude app forward.
  // A heartbeat monster (lib/cowork-heartbeat.js) can't be matched to a task: bring
  // the Claude app forward.
  if (session.client === 'cowork-heartbeat') return { kind: 'desktop-app' };

  if (session.vendor === 'claude-cowork') {
    const cowork = desktopSession(home, session.session_id, coworkStore(home));
    if (!cowork) return { status: 409, error: 'that Cowork session is no longer in the Claude app' };
    if (cowork.archived) return { status: 409, error: 'that session is archived in the Claude app' };
    if (session.state === 'blocked') return { kind: 'desktop', url: `claude://code/needs-input?session=${cowork.localId}` };
    return { kind: 'desktop-app' };
  }

  const desktop = desktopSession(home, session.session_id);
  if (desktop) {
    if (desktop.archived) return { status: 409, error: 'that session is archived in the Claude app' };
    return { kind: 'desktop', url: `claude://code/continue?session=${desktop.localId}` };
  }

  const pid = session.client_pid;
  if (!Number.isInteger(pid) || pid <= 0) return { status: 409, error: 'the zoo has no process for this session' };
  const info = ps(pid);
  if (!info) return { status: 409, error: "that session's process has exited" };
  if (!TTY_RE.test(info.tty)) return { status: 409, error: 'that session has no window to open' };

  let next = info.ppid;
  for (let i = 0; i < MAX_ANCESTORS && next > 1; i += 1) {
    const up = ps(next);
    if (!up) break;
    const app = appBundle(up.command);
    if (app) {
      return app.endsWith('/Terminal.app')
        ? { kind: 'terminal-tab', tty: `/dev/${info.tty}`, app }
        : { kind: 'app', app };
    }
    next = up.ppid;
  }
  return { status: 409, error: "couldn't find the app that session runs in" };
}

const TERMINAL_TAB_SCRIPT = [
  'on run argv',
  '  set wanted to item 1 of argv',
  '  tell application "Terminal"',
  '    repeat with w in windows',
  '      repeat with t in tabs of w',
  '        if tty of t is wanted then',
  '          set selected of t to true',
  '          set index of w to 1',
  '          activate',
  '          return "ok"',
  '        end if',
  '      end repeat',
  '    end repeat',
  '  end tell',
  '  return "not found"',
  'end run',
];

// Carries out a plan. Every argument is a validated id, tty or bundle path, passed as
// a separate argv entry: no shell, no string built into the script.
function performJump(action, exec = (cmd, args) => execFileSync(cmd, args, { encoding: 'utf8', timeout: 10000 })) {
  if (action.kind === 'desktop') {
    exec('/usr/bin/open', [action.url]);
    // The link can land in the background (anthropics/claude-code#65610); bring it forward.
    exec('/usr/bin/open', ['-b', DESKTOP_BUNDLE_ID]);
    return;
  }
  if (action.kind === 'terminal-tab') {
    let found = false;
    try {
      const args = TERMINAL_TAB_SCRIPT.flatMap((line) => ['-e', line]);
      found = String(exec('/usr/bin/osascript', [...args, action.tty])).trim() === 'ok';
    } catch {
      found = false; // no Automation permission, or Terminal refused: fall back to the app
    }
    if (!found) exec('/usr/bin/open', ['-a', action.app]);
    return;
  }
  if (action.kind === 'desktop-app') exec('/usr/bin/open', ['-b', DESKTOP_BUNDLE_ID]);
  if (action.kind === 'app') exec('/usr/bin/open', ['-a', action.app]);
}

module.exports = { planJump, performJump, desktopSession, coworkStore, parsePs, appBundle, LOCAL_ID_RE, TERMINAL_TAB_SCRIPT, DESKTOP_BUNDLE_ID };

File: src-tauri/src/jump.rs
// Port of lib/jump.js: jump to the window a session lives in. The two must agree;
// test/integration.js runs the same /jump scenarios against both servers.
//
// Desktop-app sessions open with claude://code/continue?session=<local id>, the
// Claude app's own (undocumented) route for an existing session, with the id taken
// from its session store. Terminal sessions: select the Terminal.app tab with the
// session's tty, or activate whichever app hosts it.

use serde_json::{Map, Value};
use std::path::Path;
use std::process::Command;

const DESKTOP_BUNDLE_ID: &str = "com.anthropic.claudefordesktop";
const MAX_ANCESTORS: usize = 15;

#[derive(Debug, PartialEq)]
pub enum Action {
    Desktop { url: String },
    DesktopApp,
    TerminalTab { tty: String, app: String },
    App { app: String },
}

impl Action {
    pub fn to_json(&self) -> Value {
        match self {
            Action::Desktop { url } => serde_json::json!({ "kind": "desktop", "url": url }),
            Action::DesktopApp => serde_json::json!({ "kind": "desktop-app" }),
            Action::TerminalTab { tty, app } => serde_json::json!({ "kind": "terminal-tab", "tty": tty, "app": app }),
            Action::App { app } => serde_json::json!({ "kind": "app", "app": app }),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct Refusal {
    pub status: u16,
    pub error: &'static str,
}

/// The Claude app's own check on these ids.
pub fn is_local_id(s: &str) -> bool {
    s.strip_prefix("local_")
        .is_some_and(|r| (1..=64).contains(&r.len()) && r.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'-'))
}

fn is_tty(s: &str) -> bool {
    s.strip_prefix("ttys").is_some_and(|n| (1..=4).contains(&n.len()) && n.bytes().all(|b| b.is_ascii_digit()))
}

fn subdirs(dir: &Path) -> Vec<std::path::PathBuf> {
    std::fs::read_dir(dir)
        .map(|it| it.flatten().filter(|e| e.path().is_dir()).map(|e| e.path()).collect())
        .unwrap_or_default()
}

/// (local id, archived) for the desktop session whose cliSessionId is ours.
pub fn desktop_session(home: &Path, cli_session_id: &str) -> Option<(String, bool)> {
    store_session(&home.join("Library/Application Support/Claude/claude-code-sessions"), cli_session_id)
}

/// The same lookup in the Claude app's Cowork store.
pub fn cowork_session(home: &Path, cli_session_id: &str) -> Option<(String, bool)> {
    store_session(&home.join("Library/Application Support/Claude/local-agent-mode-sessions"), cli_session_id)
}

fn store_session(root: &Path, cli_session_id: &str) -> Option<(String, bool)> {
    for org in subdirs(root).iter().flat_map(|a| subdirs(a)) {
        let Ok(entries) = std::fs::read_dir(&org) else { continue };
        for entry in entries.flatten() {
            let name = entry.file_name();
            let Some(name) = name.to_str() else { continue };
            if !name.starts_with("local_") || !name.ends_with(".json") {
                continue;
            }
            let Ok(raw) = std::fs::read_to_string(entry.path()) else { continue };
            if !raw.contains(cli_session_id) {
                continue;
            }
            let Ok(Value::Object(record)) = serde_json::from_str::<Value>(&raw) else { continue };
            let local = record.get("sessionId").and_then(Value::as_str).unwrap_or("");
            if record.get("cliSessionId").and_then(Value::as_str) == Some(cli_session_id) && is_local_id(local) {
                let archived = matches!(record.get("isArchived"), Some(Value::Bool(true)));
                return Some((local.to_string(), archived));
            }
        }
    }
    None
}

/// One `ps -o ppid=,tty=,comm=` line: (ppid, tty, command).
pub fn parse_ps(line: &str) -> Option<(i32, String, String)> {
    let line = line.trim();
    let (ppid, rest) = line.split_once(char::is_whitespace)?;
    let rest = rest.trim_start();
    let (tty, command) = rest.split_once(char::is_whitespace)?;
    let command = command.trim();
    (!command.is_empty()).then(|| Some((ppid.parse().ok()?, tty.to_string(), command.to_string())))?
}

pub fn ps_info(pid: i32) -> Option<(i32, String, String)> {
    let out = Command::new("/bin/ps").args(["-o", "ppid=,tty=,comm=", "-p", &pid.to_string()]).output().ok()?;
    parse_ps(&String::from_utf8_lossy(&out.stdout))
}

/// "/Applications/X.app/Contents/MacOS/X" -> "/Applications/X.app", skipping Claude
/// Code's own claude.app bundle.
pub fn app_bundle(command: &str) -> Option<&str> {
    let end = command.find(".app/Contents/MacOS/")? + ".app".len();
    let bundle = &command[..end];
    (!bundle.ends_with("/claude.app")).then_some(bundle)
}

pub fn plan(session: &Map<String, Value>, home: &Path, ps: &dyn Fn(i32) -> Option<(i32, String, String)>) -> Result<Action, Refusal> {
    let Some(id) = session.get("session_id").and_then(Value::as_str) else {
        return Err(Refusal { status: 404, error: "no such session" });
    };
    // A Cowork heartbeat monster can't be matched to a task: bring the app forward.
    if session.get("client").and_then(Value::as_str) == Some("cowork-heartbeat") {
        return Ok(Action::DesktopApp);
    }
    // Cowork: the "waiting for you" link opens a waiting Cowork session; any other
    // Cowork state just brings the Claude app forward.
    if session.get("vendor").and_then(Value::as_str) == Some("claude-cowork") {
        let Some((local, archived)) = cowork_session(home, id) else {
            return Err(Refusal { status: 409, error: "that Cowork session is no longer in the Claude app" });
        };
        if archived {
            return Err(Refusal { status: 409, error: "that session is archived in the Claude app" });
        }
        if session.get("state").and_then(Value::as_str) == Some("blocked") {
            return Ok(Action::Desktop { url: format!("claude://code/needs-input?session={local}") });
        }
        return Ok(Action::DesktopApp);
    }
    if let Some((local, archived)) = desktop_session(home, id) {
        if archived {
            return Err(Refusal { status: 409, error: "that session is archived in the Claude app" });
        }
        return Ok(Action::Desktop { url: format!("claude://code/continue?session={local}") });
    }
    let pid = session.get("client_pid").and_then(Value::as_f64).filter(|p| p.fract() == 0.0 && *p > 0.0 && *p <= i32::MAX as f64);
    let Some(pid) = pid.map(|p| p as i32) else {
        return Err(Refusal { status: 409, error: "the zoo has no process for this session" });
    };
    let Some((mut next, tty, _)) = ps(pid) else {
        return Err(Refusal { status: 409, error: "that session's process has exited" });
    };
    if !is_tty(&tty) {
        return Err(Refusal { status: 409, error: "that session has no window to open" });
    }
    for _ in 0..MAX_ANCESTORS {
        if next <= 1 {
            break;
        }
        let Some((ppid, _, command)) = ps(next) else { break };
        if let Some(app) = app_bundle(&command) {
            return Ok(if app.ends_with("/Terminal.app") {
                Action::TerminalTab { tty: format!("/dev/{tty}"), app: app.to_string() }
            } else {
                Action::App { app: app.to_string() }
            });
        }
        next = ppid;
    }
    Err(Refusal { status: 409, error: "couldn't find the app that session runs in" })
}

const TERMINAL_TAB_SCRIPT: [&str; 16] = [
    "on run argv",
    "  set wanted to item 1 of argv",
    "  tell application \"Terminal\"",
    "    repeat with w in windows",
    "      repeat with t in tabs of w",
    "        if tty of t is wanted then",
    "          set selected of t to true",
    "          set index of w to 1",
    "          activate",
    "          return \"ok\"",
    "        end if",
    "      end repeat",
    "    end repeat",
    "  end tell",
    "  return \"not found\"",
    "end run",
];

fn run(cmd: &str, args: &[&str]) -> Option<String> {
    Command::new(cmd).args(args).output().ok().map(|o| String::from_utf8_lossy(&o.stdout).to_string())
}

/// Carries out a plan; every value is a separate argv entry, never a shell string.
pub fn perform(action: &Action) {
    match action {
        Action::Desktop { url } => {
            run("/usr/bin/open", &[url]);
            // The link can land in the background (anthropics/claude-code#65610).
            run("/usr/bin/open", &["-b", DESKTOP_BUNDLE_ID]);
        }
        Action::TerminalTab { tty, app } => {
            let mut args: Vec<&str> = TERMINAL_TAB_SCRIPT.iter().flat_map(|l| ["-e", *l]).collect();
            args.push(tty);
            let found = run("/usr/bin/osascript", &args).is_some_and(|out| out.trim() == "ok");
            if !found {
                run("/usr/bin/open", &["-a", app]);
            }
        }
        Action::DesktopApp => {
            run("/usr/bin/open", &["-b", DESKTOP_BUNDLE_ID]);
        }
        Action::App { app } => {
            run("/usr/bin/open", &["-a", app]);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn table(rows: &'static [(i32, i32, &'static str, &'static str)]) -> impl Fn(i32) -> Option<(i32, String, String)> {
        move |pid| rows.iter().find(|r| r.0 == pid).map(|r| (r.1, r.2.to_string(), r.3.to_string()))
    }

    fn session(pid: Option<i32>) -> Map<String, Value> {
        let mut m = Map::new();
        m.insert("session_id".into(), "cli-1".into());
        if let Some(p) = pid {
            m.insert("client_pid".into(), p.into());
        }
        m
    }

    #[test]
    fn ids_ttys_and_ps_lines() {
        assert!(is_local_id("local_abc-123"));
        assert!(!is_local_id("local_bad;rm"));
        assert!(!is_local_id("local_"));
        assert!(is_tty("ttys003"));
        assert!(!is_tty("??"));
        assert_eq!(
            parse_ps("  123 ttys004 /Applications/Some App.app/Contents/MacOS/Some App\n"),
            Some((123, "ttys004".into(), "/Applications/Some App.app/Contents/MacOS/Some App".into()))
        );
        assert_eq!(parse_ps(""), None);
        assert_eq!(app_bundle("/Applications/Some App.app/Contents/MacOS/Some App"), Some("/Applications/Some App.app"));
        assert_eq!(app_bundle("/x/claude-code/claude.app/Contents/MacOS/claude"), None);
        assert_eq!(app_bundle("-zsh"), None);
    }

    #[test]
    fn terminal_tab_other_app_and_refusals() {
        let home = std::env::temp_dir().join(format!("zoo-jump-rs-{}", std::process::id()));
        let term = table(&[
            (500, 400, "ttys003", "claude"),
            (400, 300, "ttys003", "-zsh"),
            (300, 1, "??", "/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal"),
        ]);
        assert_eq!(
            plan(&session(Some(500)), &home, &term),
            Ok(Action::TerminalTab { tty: "/dev/ttys003".into(), app: "/System/Applications/Utilities/Terminal.app".into() })
        );
        let ghostty = table(&[(500, 400, "ttys001", "claude"), (400, 1, "??", "/Applications/Ghostty.app/Contents/MacOS/ghostty")]);
        assert_eq!(plan(&session(Some(500)), &home, &ghostty), Ok(Action::App { app: "/Applications/Ghostty.app".into() }));
        assert_eq!(plan(&session(None), &home, &term).unwrap_err().status, 409);
        let headless = table(&[(9, 1, "??", "claude")]);
        assert_eq!(plan(&session(Some(9)), &home, &headless).unwrap_err().error, "that session has no window to open");
    }

    #[test]
    fn cowork_sessions_open_while_waiting_and_otherwise_bring_the_app_forward() {
        let home = std::env::temp_dir().join(format!("zoo-jump-rs-cowork-{}", std::process::id()));
        let dir = home.join("Library/Application Support/Claude/local-agent-mode-sessions/acct/org");
        std::fs::create_dir_all(&dir).unwrap();
        std::fs::write(dir.join("local_cw1.json"), r#"{"sessionId":"local_cw1","cliSessionId":"cli-1","isArchived":false}"#).unwrap();
        let none = |_| None;
        let mut s = session(None);
        s.insert("vendor".into(), "claude-cowork".into());
        s.insert("state".into(), "blocked".into());
        assert_eq!(plan(&s, &home, &none), Ok(Action::Desktop { url: "claude://code/needs-input?session=local_cw1".into() }));
        s.insert("state".into(), "done".into());
        assert_eq!(plan(&s, &home, &none), Ok(Action::DesktopApp));
        std::fs::remove_file(dir.join("local_cw1.json")).unwrap();
        assert_eq!(plan(&s, &home, &none).unwrap_err().status, 409);
        let _ = std::fs::remove_dir_all(&home);
    }

    #[test]
    fn cowork_heartbeat_monsters_bring_the_app_forward() {
        let home = std::env::temp_dir().join(format!("zoo-jump-rs-hb-{}", std::process::id()));
        let mut s = session(None);
        s.insert("vendor".into(), "claude-cowork".into());
        s.insert("client".into(), "cowork-heartbeat".into());
        s.insert("state".into(), "working".into());
        let none = |_| None;
        assert_eq!(plan(&s, &home, &none), Ok(Action::DesktopApp));
    }

    #[test]
    fn desktop_sessions_are_found_in_the_app_store() {
        let home = std::env::temp_dir().join(format!("zoo-jump-rs-desk-{}", std::process::id()));
        let dir = home.join("Library/Application Support/Claude/claude-code-sessions/acct/org");
        std::fs::create_dir_all(&dir).unwrap();
        std::fs::write(dir.join("local_abc.json"), r#"{"sessionId":"local_abc","cliSessionId":"cli-1","isArchived":false}"#).unwrap();
        let none = |_| None;
        assert_eq!(plan(&session(None), &home, &none), Ok(Action::Desktop { url: "claude://code/continue?session=local_abc".into() }));
        std::fs::write(dir.join("local_abc.json"), r#"{"sessionId":"local_abc","cliSessionId":"cli-1","isArchived":true}"#).unwrap();
        assert_eq!(plan(&session(None), &home, &none).unwrap_err().status, 409);
        let _ = std::fs::remove_dir_all(&home);
    }
}
